tuta-sdk/rust/sdk/src/entities/entity_facade.rs

```diff
@@ -41,15 +62,238 @@ pub trait EntityFacade: Send + Sync {
+			ValueType::CompressedString => unimplemented!("compressed string"),
+		}
+	}
+
+	fn encrypt_and_map_to_literal_with_iv(
+		&self,
+		type_model: &TypeModel,
+		instance: &ParsedEntity,
+		sk: Option<GenericAesKey>,
+		random_iv_provider: IvProvider,
+	) -> Result<ParsedEntity, ApiCallError> {
+		if type_model.marked_encrypted() && sk.is_none() {
+			return Err(ApiCallError::InternalSdkError {
+				error_message: format!(
+					"Encrypting {}/{} requires a session key!",
+					type_model.app, type_model.name
+				),
+			});
+		}
+		let mut encrypted = ParsedEntity::new();
+
+		for (key, model_value) in &type_model.values {
+			let instance_value = instance
+				.get(&key.to_string())
+				.unwrap_or_else(|| panic!("Can not find key: {key} in instance: {instance:?}"));
+
+			let encrypted_value: ElementValue;
+
+			if Self::should_restore_default_value(model_value, instance_value, instance, key) {
+				// restore the default encrypted value because it has not changed
+				// note: this branch must be checked *before* the one which reuses IVs as this one checks
+				// the length.
+				encrypted_value = ElementValue::String("".to_string());
+			} else if model_value.encrypted
+				&& model_value.is_final
+				&& Self::get_final_iv_for_key(instance, key).is_some()
+			{
+				let final_iv = Iv::from_bytes(
+					Self::get_final_iv_for_key(instance, key)
+						.unwrap()
+						.assert_bytes()
+						.as_slice(),
+				)
+				.map_err(|err| ApiCallError::internal(format!("iv of illegal size {:?}", err)))?;
+				encrypted_value = Self::encrypt_value(
+					key,
+					model_value,
+					instance_value,
+					&sk,
+					IvProvider::Fixed(final_iv),
+				)?
+			} else {
+				encrypted_value = Self::encrypt_value(
+					key,
+					model_value,
+					instance_value,
+					&sk,
+					random_iv_provider.clone(),
+				)?
+			}
+			encrypted.insert(key.to_string(), encrypted_value);
+		}
+
+		if type_model.element_type == ElementType::Aggregated && !encrypted.contains_key("_id") {
+			let randomizer = RandomizerFacade::from_core(rand_core::OsRng);
+			let new_id = randomizer.generate_random_array::<4>();
+
+			encrypted.insert(
+				String::from("_id"),
+				ElementValue::String(BASE64_URL_SAFE_NO_PAD.encode(BASE64_STANDARD.encode(new_id))),
+			);
+		}
+
+		for (association_name, association) in &type_model.associations {
+			if let AssociationType::Aggregation = association.association_type {
+				let dependency = association.dependency.unwrap_or(type_model.app);
+				let aggregated_type_model = self
+					.type_model_provider
+					.get_type_model(dependency, association.ref_type)
+					.unwrap_or_else(|| {
+						panic!(
+							"Cannot find type model for: {:?}",
+							(dependency, association.ref_type)
+						)
+					});
+				let aggregation = association;
+				let instance_association = instance.get(&association_name.to_string()).unwrap();
+				if aggregation.cardinality == Cardinality::ZeroOrOne
```

Comment:
I would feel slightly better if this `if` was a `match` so that we explicitly mention all the cases. It might also be more readable.

Reply:
🙇‍♂️